core/domain/learner_playlist_services_test.py

```diff
@@ -100,162 +101,128 @@ def _get_all_learner_playlist_collection_ids(self, user_id):
         self.assertEqual(
             self._get_all_learner_playlist_exp_ids(self.user_id), [])
 
-        # Add an exploration to the learner playlist.
+        learner_playlist_services.remove_exploration_from_learner_playlist(
+            self.user_id, 'SUBSCRIBED_ID')
+
+        # Test adding the exploration_id if it is already in
+        # learner_playlist.exploration_ids
+        # Add two explorations to the learner playlist.
         learner_progress_services.add_exp_to_learner_playlist(
             self.user_id, self.EXP_ID_0)
-        self.assertEqual(
-            self._get_all_learner_playlist_exp_ids(self.user_id),
-            [self.EXP_ID_0])
-
-        # Add another exploration.
         learner_progress_services.add_exp_to_learner_playlist(
             self.user_id, self.EXP_ID_1)
         self.assertEqual(
             self._get_all_learner_playlist_exp_ids(
                 self.user_id), [self.EXP_ID_0, self.EXP_ID_1])
 
-        # Add the first exploration in a different position.
+        # Add the first exploration to the second position.
         learner_progress_services.add_exp_to_learner_playlist(
             self.user_id, self.EXP_ID_0,
             position_to_be_inserted=1)
         self.assertEqual(
             self._get_all_learner_playlist_exp_ids(
                 self.user_id), [self.EXP_ID_1, self.EXP_ID_0])
 
-        # Add the first exploration in the first position.
-        learner_progress_services.add_exp_to_learner_playlist(
-            self.user_id, self.EXP_ID_0,
-            position_to_be_inserted=0)
-        self.assertEqual(
-            self._get_all_learner_playlist_exp_ids(
-                self.user_id), [self.EXP_ID_0, self.EXP_ID_1])
-
-        # Add an exploration to the in progress list of the user.
-        state_name = 'state_name'
-        version = 1
-        learner_progress_services.mark_exploration_as_incomplete(
-            self.user_id, self.EXP_ID_3, state_name, version)
-
-        # Test that the exploration added to the in progress list doesn't get
-        # added to the learner playlist.
-        learner_progress_services.add_exp_to_learner_playlist(
-            self.user_id, self.EXP_ID_3)
-        self.assertEqual(
-            self._get_all_learner_playlist_exp_ids(
-                self.user_id), [self.EXP_ID_0, self.EXP_ID_1])
-
-        # Empty the learner playlist.
-        learner_playlist_services.remove_exploration_from_learner_playlist(
-            self.user_id, self.EXP_ID_0)
-        learner_playlist_services.remove_exploration_from_learner_playlist(
-            self.user_id, self.EXP_ID_1)
-        self.assertEqual(
-            self._get_all_learner_playlist_exp_ids(
-                self.user_id), [])
-
-        # Test that the length of the learner playlist doesn't exceed
-        # MAX_LEARNER_PLAYLIST_ACTIVITY_COUNT.
-        # List of explorations to be added.
+    def test_nunmber_of_explorations_cannot_exceed_max(self):
```

Comment:
Split this into two test:

- `test_new_exp_is_not_added_at_list_end_after_list_size_is_max`
- `test_new_exp_is_not_added_at_random_index_after_list_size_is_max`

Reply:
I originally had this, but @seanlip said I could combine the 2 tests (1 at list end and 1 at random index) into one test. Which should I do?

Reply:
Technically both options are fine as long as the test name is really clear. That's because the position_to_be_inserted does not really make any difference (so it's a bit of a red herring). If it did, separate tests would be better.

So I think what you have now is fine.

Reply:
Okay thanks!